doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,331 @@
+- Functions that are narrow in scope, with very few public use-cases,
+  will be removed.  See, e.g.
+  ``real_if_close`` (`gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
+  These will have to be identified manually and by issue triage.
+
+New namespaces are introduced for warnings/exceptions (``np.exceptions``) and
+for dtype-related functionality (``np.types``). NumPy 2.0 is a good opportunity
+to populate these submodules from the main namespace.
+
+Functionality that is widely used but has a preferred alternative may either be
+deprecated (with the deprecation message pointing out what to use instead) or
+be hidden by not including it in ``__dir__``. In case of hiding, a ``..
+legacy::`` directory may be used to mark such functionality in the
+documentation.
+
+A test will be added to ensure limited future growth of all namespaces; i.e.,
+every new entry will need to be explicitly added to an allow-list.
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces.
+Details TBD, something like:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.ma
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.math
+    numpy.lib.stride_tricks
+    numpy.rec
+    numpy.types
+
+    # Legacy (prefer not to use). Third grouping in the reference guide.
+    numpy.char
+    numpy.distutils
+    numpy.matrixlib  # or deprecate?
+
+    # To remove
+    numpy.compat
+    numpy.core  # rename to _core
+    numpy.doc
+    numpy.matlib
+    numpy.version
+
+    # To clean out or somehow deal with: everything in `numpy.lib`
+
+.. note::
+
+    TBD: will we preserve ``np.lib`` or not? It only has a couple of unique
```

Comment:
Agreed, but it would help here to know where would things like `arraysetops` and `nanfunctions` would live (I realize this is not so important from a user perspective, since the functions are in the top namespace, but it is probably also good to think from the developer perspective).

Reply:
I'd say that they have to move somewhere, which is a single `git mv` plus use of `git-blame-ignore-revs` to not mess up the usefulness of `git blame`. Where it ends up exactly is less important to me, at least from the perspective of this NEP. It could be:
- `numpy/_lib/nanfunctions.py`
- `numpy/lib/_nanfunctions.py`
- or something more different than adding a single underscore